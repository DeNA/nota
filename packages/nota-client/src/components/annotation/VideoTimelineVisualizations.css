.vis-container {
  background-color: rgba(50, 50, 50, 0.9);
  display: flex;
  align-items: center;
  justify-content: center;
  position: relative;
  min-height: 60px;
  max-height: 60px;
  width: 100%;
}

.vis-container .rv-crosshair__inner {
  bottom: 0 !important;
  top: auto !important;
}
