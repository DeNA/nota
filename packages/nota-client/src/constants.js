export const IMAGE_STATUS_ONGOING = 0;
export const IMAGE_STATUS_DONE = 1;

export const ANNOTATION_LINE_WIDTH = 2;
export const ANNOTATION_DRAGLINE_WIDTH = 30;

export const GROUP_ADMIN = "admin";
export const GROUP_USER = "user";

export const GROUP_OPTIONS = [
  { key: "user", text: "User", value: "user" },
  { key: "admin", text: "Admin", value: "admin" }
];

export const FOLDER_STATUS = {
  0: "CREATING",
  50: "ERROR",
  100: "READY",
  200: "HIDDEN"
};

/** Flow Options */
export const OPTION_NOOP = "none";
export const OPTION_FOCUS = "focus";
export const OPTION_POPUP_FOCUS = "popup_focus";
export const OPTION_LABEL_SHOW_OFF = "label_show_off";
export const OPTION_LABEL_SHOW_HOVER = "label_show_hover";
export const OPTION_LABEL_SHOW_ALWAYS = "label_show_always";

export const ACTION_SELECT_ANNOTATION = "select_annotation";
export const ACTION_SELECT_ANNOTATION_USER = "select_annotation_user";
export const ACTION_CREATE_ANNOTATION = "create_annotation";

export const MEDIA_TYPE = {
  IMAGE: "IMAGE",
  VIDEO: "VIDEO"
};

export const DEFAULT_ANNOTATION_COLOR = "rgba(200,255,100,1)";
export const DEFAULT_SELECTED_ANNOTATION_COLOR = "rgba(200,255,100,1)";
