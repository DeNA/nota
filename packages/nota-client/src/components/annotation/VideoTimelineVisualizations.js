import React, { useContext, useEffect, useState } from "react";
import useDimensions from "react-cool-dimensions";
import { useTranslation } from "react-i18next";
import { Crosshair, LineSeries, XYPlot } from "react-vis";
import "react-vis/dist/style.css";
import { getVis } from "../../lib/binaryVis";
import Loading from "../Loading";
import { videoControlsContext } from "./videoControls";
import "./VideoTimelineVisualizations.css";

const DEFAULT_COLOR = "red";

function VideoTimelineVisualizations({
  projectId,
  taskId,
  taskItemId,
  timelineVis = [],
  min,
  max
}) {
  const { t } = useTranslation();
  const hasVis = timelineVis.length > 0;
  const { observe, width, height } = useDimensions();
  const [tooltip, setTooltip] = useState(null);
  const [visualizations, setVisualizations] = useState({});
  const [loading, setLoading] = useState(false);
  const {
    videoControls: { setTime, getDuration }
  } = useContext(videoControlsContext);

  useEffect(() => {
    if (hasVis && projectId && taskId && taskItemId) {
      setLoading(true);
      getVis(projectId, taskId, taskItemId)
        .then(visualizations => {
          setVisualizations(visualizations);
        })
        .catch(error => {
          console.log(error);
        })
        .finally(() => {
          setLoading(false);
        });
    }
  }, [hasVis, projectId, taskId, taskItemId]);

  const lines = timelineVis
    .map(visOptions => {
      const vis = visualizations[visOptions.id];

      return vis
        ? {
            id: visOptions.id,
            label: vis.label || visOptions.label || visOptions.id,
            color: vis.color || visOptions.color || DEFAULT_COLOR,
            data: vis.values.map(([time, value]) => {
              return {
                x: time,
                y: value
              };
            })
          }
        : null;
    })
    .filter(line => line !== null);

  const handleOnNearestX = function({ x, y }, { index }) {
    const tooltip = lines
      .map(line => {
        const found = line.data.find(data => data.x === x);

        return found
          ? {
              label: line.label,
              id: line.id,
              y: found.y,
              x: found.x
            }
          : null;
      })
      .filter(data => data !== null);

    setTooltip(tooltip);
  };

  const handleOnClick = function() {
    if (tooltip?.[0]?.x) {
      setTime(tooltip[0].x);
    }
  };

  if (!hasVis) {
    return null;
  }

  const showGraph = !loading && lines.length > 0;

  return (
    <div ref={observe} className="vis-container">
      <Loading loading={loading} />
      {showGraph ? (
        <div className="timeline-chart" data-testid="timeline-chart">
          <XYPlot
            width={width}
            height={height}
            margin={{ left: 3, right: 3, top: 5, bottom: 5 }}
            xDomain={[0, getDuration()]}
            yDomain={[min, max]}
            onClick={handleOnClick}
          >
            {lines.map((line, i) => (
              <LineSeries
                key={line.id}
                className={line.id}
                color={line.color}
                data={line.data}
                onNearestX={i === 0 ? handleOnNearestX : null}
              />
            ))}
            <Crosshair
              values={tooltip}
              itemsFormat={values => {
                return values.map(d => ({
                  title: d.label,
                  value: d.y
                }));
              }}
              titleFormat={values => {
                return {
                  title: "ms",
                  value: values[0].x
                };
              }}
            />
          </XYPlot>
        </div>
      ) : (
        <div>{t("vis-not-found")}</div>
      )}
    </div>
  );
}

export default VideoTimelineVisualizations;
