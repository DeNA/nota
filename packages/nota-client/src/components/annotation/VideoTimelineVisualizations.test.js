jest.mock("../../lib/binaryVis");
import { ResizeObserver, ResizeObserverEntry } from "@juggle/resize-observer";
import { act, render, screen } from "@testing-library/react";
import React from "react";
import { ALL_PRESET } from "../../constants";
import VideoControlsProvider from "./videoControls";
import VideoTimelineVisualizations from "./VideoTimelineVisualizations";

beforeEach(() => {
  window.ResizeObserver = ResizeObserver;
  window.ResizeObserverEntry = ResizeObserverEntry;
});

describe("VideoTimeLineVisualizations", () => {
  test("render nothing without data", async () => {
    let tree;
    const ALL = { id: ALL_PRESET, label: "All", vis: [] };

    act(() => {
      tree = render(
        <VideoControlsProvider>
          <VideoTimelineVisualizations
            projectId={1}
            taskId={1}
            taskItemId={1}
            timelineVis={[]}
            min={0}
            max={1}
            timelineVisPresets={[ALL]}
            selectedPreset={ALL}
            changeTimelineVisState={() => {}}
          />
        </VideoControlsProvider>
      );
    });

    expect(tree.asFragment()).toMatchSnapshot();
  });

  test("render timeline correctly", async () => {
    let tree;
    const ALL = {
      id: ALL_PRESET,
      label: "All",
      vis: ["test_timeline", "test_timeline_2", "test_timeline_not_existent"]
    };

    act(() => {
      tree = render(
        <VideoControlsProvider>
          <VideoTimelineVisualizations
            projectId={1}
            taskId={1}
            taskItemId={1}
            timelineVis={[
              {
                id: "test_timeline",
                label: "LINE_A"
              },
              {
                id: "test_timeline_2",
                color: "green"
              },
              {
                id: "test_timeline_not_existent",
                label: "LINE_C"
              }
            ]}
            min={0}
            max={1}
            timelineVisPresets={[ALL]}
            selectedPreset={ALL}
            changeTimelineVisState={() => {}}
          />
        </VideoControlsProvider>
      );
    });

    expect(tree.asFragment()).toMatchSnapshot();

    await act(async () => {});
    await screen.findByTestId("timeline-chart", {}, {});
    expect(tree.asFragment()).toMatchSnapshot();
  });

  test("render message when no matching vis", async () => {
    let tree;
    const ALL = { id: ALL_PRESET, label: "All", vis: ["LINE_C"] };

    act(() => {
      tree = render(
        <VideoControlsProvider>
          <VideoTimelineVisualizations
            projectId={1}
            taskId={1}
            taskItemId={1}
            timelineVis={[
              {
                id: "test_timeline_not_existent",
                label: "LINE_C"
              }
            ]}
            min={0}
            max={1}
            timelineVisPresets={[ALL]}
            selectedPreset={ALL}
            changeTimelineVisState={() => {}}
          />
        </VideoControlsProvider>
      );
    });

    await act(async () => {});
    expect(tree.asFragment()).toMatchSnapshot();
  });
});
