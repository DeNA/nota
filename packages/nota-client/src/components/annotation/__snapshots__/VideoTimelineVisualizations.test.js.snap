// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoTimeLineVisualizations render message when no matching vis 1`] = `
<DocumentFragment>
  <div
    class="vis-container"
  >
    <div>
      vis-not-found
    </div>
    <div
      class="toggle-controls"
    >
      <span
        aria-hidden="true"
        class="oi oi-cog"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`VideoTimeLineVisualizations render nothing without data 1`] = `<DocumentFragment />`;

exports[`VideoTimeLineVisualizations render timeline correctly 1`] = `
<DocumentFragment>
  <div
    class="vis-container"
  >
    <div
      class="d-flex justify-content-center align-items-center"
      style="height: 100%; background-color: rgba(0, 0, 0, 0.1); position: absolute; top: 0px; bottom: 0px; left: 0px; right: 0px; z-index: 10000;"
    >
      <div
        class="spinner-border"
      />
    </div>
    <div>
      vis-not-found
    </div>
    <div
      class="toggle-controls"
    >
      <span
        aria-hidden="true"
        class="oi oi-cog"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`VideoTimeLineVisualizations render timeline correctly 2`] = `
<DocumentFragment>
  <div
    class="vis-container"
  >
    <div
      class="timeline-chart"
      data-testid="timeline-chart"
    >
      <div
        class="timeline-progress"
        style="left: -1px;"
      />
      <div
        class="rv-xy-plot "
        style="width: 0px; height: 0px;"
      >
        <svg
          class="rv-xy-plot__inner"
          height="0"
          width="0"
        >
          <path
            class="rv-xy-plot__series rv-xy-plot__series--line test_timeline"
            d="M-6,-10Z"
            style="opacity: 1; stroke: red;"
            transform="translate(3,5)"
          />
          <path
            class="rv-xy-plot__series rv-xy-plot__series--line test_timeline_2"
            d="M-6,0Z"
            style="opacity: 1; stroke: green;"
            transform="translate(3,5)"
          />
        </svg>
      </div>
    </div>
    <div
      class="toggle-controls"
    >
      <span
        aria-hidden="true"
        class="oi oi-cog"
      />
    </div>
  </div>
</DocumentFragment>
`;
